src/instrument_meta.py

```diff
@@ -0,0 +1,307 @@
+
+    def __post_init__(self):
+        self.method = db_utils.validate_method(self.request_dict.get('method'))
+        self.params = self.request_dict.get('params')
+        self.body = self.request_dict.get('body')
+        self.filters = None
+        self.ordering = None
+        self.record_limit = None
+
+        if self.method == db_utils.HTTP_PUT:
+            self.instrument_meta = get_instrument_meta_from_body(self.body)
+        else:
+            if isinstance(self.params, dict):
+                self.filters = construct_filters(self.params.get('filters'))
+                self.ordering = self.params.get('ordering')
+                self.record_limit = self.params.get('record_limit')
+
+                if not type(self.record_limit) == int or self.record_limit <= 0:
+                    self.record_limit = None
+            else:
+                self.filters = None
+                self.ordering = None
+                self.record_limit = None
+            
+
+    def failed_request(self, error_msg):
+        return DbActionResponse(
+            request=self.request_dict,
+            success=False,
+            message='Failed instrument meta request',
+            details=None, 
+            errors=error_msg
+        )
+
+    def submit(self):
+        if self.method == db_utils.HTTP_GET:
+            return self.get_instrument_metas()
+        elif self.method == db_utils.HTTP_PUT:
+            try:
+                return self.put_instrument_meta()
+            except Exception as err:
+                error_msg = 'Failed to insert instrument meta record -'\
+                    f' err: {err}'
+                print(f'Submit PUT instrument meta error: {error_msg}')
+                return self.failed_request(error_msg)
+
+    def put_instrument_meta(self):
+        session = stm.get_session()
+
+        insert_stmt = insert(im).values(
+            name = self.instrument_meta.name,
+            num_channels = self.instrument_meta.num_channels,
+            scan_angle = self.instrument_meta.scan_angle,
+            created_at = datetime.utcnow(),
+            updated_at = None
+        ).returning(im)
+        print(f'insert stmt: {insert_stmt}')
+
+        time_now = datetime.utcnow()
+
+        do_update_stmt = insert_stmt.on_conflict_do_update(
+            constraint='unique_instrument_meta',
+            set_=dict(
+                num_channels = self.instrument_meta.num_channels,
+                scan_angle = self.instrument_meta.scan_angle,
+                updated_at = time_now
+            )
+        )
+
+        print(f'do update stmt: {do_update_stmt}')
+
+        try:
+            result = session.execute(do_update_stmt)
+            session.flush()
+            result_row = result.fetchone()
+            action = db_utils.INSERT
+            if result_row.updated_at is not None:
+                action = db_utils.UPDATE
+
+            session.commit()
+            session.close()
+        except Exception as err:
```

Comment:
add specific exception?

Reply:
same as the other exceptions mentioned, not clearly a single exception but we don't want to proceed if it fails, and we definitely want to close the session.